graphql/schemabuilder/pagination.go

```diff
@@ -0,0 +1,530 @@
+			}
+			return nil, fmt.Errorf("error resolving edges in connection")
+		},
+		Type:           edgesSliceType,
+		ParseArguments: nilParseArguments,
+	}
+
+	fieldMap["edges"] = edgesSliceField
+
+	pageInfoType, _ := reflect.TypeOf(Connection{}).FieldByName("PageInfo")
+	pageInfoField, err := sb.buildField(pageInfoType)
+	if err != nil {
+		return nil, err
+	}
+	fieldMap["pageInfo"] = pageInfoField
+
+	retObject := &graphql.NonNull{Type: &graphql.Object{Name: "Connection", Description: "", Fields: fieldMap}}
+	return retObject, nil
+}
+
+// EdgesToReturn returns the slice of edges by appyling the pagination arguments. It also returns
+// the hasNextPage and hasPrevPage values respectively. The behavior is expected to conform to the
+// Relay Cursor spec: https://facebook.github.io/relay/graphql/connections.htm#EdgesToReturn()
+func EdgesToReturn(allEdges []Edge, before *string, after *string, first *int64, last *int64) ([]Edge, bool, bool, error) {
+
+	edges, elemsAfter, elemsBefore := applyCursorsToAllEdges(allEdges, before, after)
+
+	prevPage := false
+	nextPage := false
+
+	if first != nil {
+		if *first < 0 {
+			return nil, nextPage, prevPage, graphql.NewClientError("first should be a non-negative integer")
+		}
+		if len(edges) > int(*first) {
+			edges = edges[:int(*first)]
+			nextPage = true
+		}
+	}
+	if before != nil {
+		if elemsAfter {
+			nextPage = true
+		}
+	}
+
+	if last != nil {
+		if *last < 0 {
+			return nil, nextPage, prevPage, graphql.NewClientError("last should be a non-negative integer")
+		}
+		if len(edges) > int(*last) {
+			edges = edges[len(edges)-int(*last):]
+			prevPage = true
+		}
+	}
+	if after != nil {
+		if elemsBefore {
+			prevPage = true
+		}
+	}
+
+	return edges, nextPage, prevPage, nil
+
+}
+
+// getCursorIndex returns the index corresponding to the cursor in the slice.
+func getCursorIndex(edges []Edge, cursor string) int {
+
+	for i, val := range edges {
+		if val.Cursor == cursor {
+			return i
+		}
+	}
+	return -1
+}
+
+// applyCursorsToAllEdges returns the slice of edges after applying the after and before arguments.
+// It also implements part of the hasNextPage and hasPrevPage algorithm by returning if there are
+// elements after or before the arguments.
+func applyCursorsToAllEdges(allEdges []Edge, before *string, after *string) ([]Edge, bool, bool) {
+
+	edges := allEdges
+
+	elemsAfter := false
+	elemsBefore := false
+
+	if after != nil {
+		i := getCursorIndex(edges, *after)
+		if i != -1 {
+			edges = edges[i+1:]
+			if i != 0 {
+				elemsBefore = true
+			}
+		}
+
+	}
+	if before != nil {
+		i := getCursorIndex(edges, *before)
+		if i != -1 {
+			edges = edges[:i]
+			if i != len(allEdges)-1 {
+				elemsAfter = true
+			}
+		}
+
+	}
+
+	return edges, elemsAfter, elemsBefore
+
+}
+
+// getConnection applies the ConnectionArgs to nodes and returns the result in a wrapped Connection
+// type.
+func getConnection(key string, nodes []interface{}, args ConnectionArgs) (interface{}, error) {
```

Comment:
why are we returning an interface? can't you return a Connection?

Reply:
to maintain some uniformity with the code in reflect?

Reply:
but yes, Connection makes more sense